lib/logstash/inputs/s3.rb

```diff
@@ -147,49 +150,60 @@ def process_new(queue, since=nil)
         since = sincedb_read()
     end
 
-    objects = list_new(since)
-    objects.each do |k|
-      @logger.debug("S3 input processing", :bucket => @bucket, :key => k)
-      lastmod = @s3bucket.objects[k].last_modified
-      process_log(queue, k)
-      sincedb_write(lastmod)
-    end
+    fetch_new_and_process(queue, since)
 
   end # def process_new
 
   private
-  def list_new(since=nil)
+  def fetch_new_and_process(queue, since=nil)
 
     if since.nil?
       since = Time.new(0)
     end
 
-    objects = {}
+    objects = []
     @s3bucket.objects.with_prefix(@prefix).each do |log|
-      if log.last_modified > since
-        objects[log.key] = log.last_modified
+      # We skip already processed objects.
+      # Zero length objects have nothing to process.
+      # They can represent "directories". So we skip them as well.
+      if log.last_modified > since && log.content_length > 0
+        objects << log
+        if objects.count >= @batch_size
+          process_s3_objects(queue, objects)
+          objects.clear
+        end
       end
     end
+    if not objects.empty?
+      process_s3_objects(queue, objects)
+    end
 
-    return sorted_objects = objects.keys.sort {|a,b| objects[a] <=> objects[b]}
```

Comment:
It seems like you've eliminated the sort step. In cases where alphabetical order matches chronological order by last_modified time, this will work (since the S3 API always returns results in alphabetical order) but if not this creates a problem because the sincedb assumes that objects are always handled in the same order.